#RNA seq

<sub>
motivation
</sub>

Today, we will visualise mapped reads, count reads, visualise counts, transform counts to more practical units and perform a differential expression analysis.

## Previously in Sequence a genome...

- remote connection and file transfer `ssh`, `scp` to server `<username>@prd.vital-it.ch`
- basic commands `pwd`, `ls`, `cd`, `cp`, `mv`, `mkdir` and finally `man` if you forgot anything
- text processing tools `cat`, `head`, `tail`, `less` and `vi`
- operations on cluster `module add`, `bsub` and `bjobs`
- `fastqc`, `kallisto` and `R` prepared to use (on vital-it or your computer)
- RNAseq theory
- quality control of reads using `fastqc`
- building a reference and mapping reads using `kallisto`

***

## Input data and preparations

<sub>
The starting point for an RNA-Seq experiment is a set of N RNA samples, typically associated with a variety of treatment conditions. Each sample is sequenced, short reads are mapped to the appropriate genome, and the number of reads mapped to each genomic feature of interest is recorded. 
The set of gene-wise counts for sample j makes up the expression profile or library for that sample. The expected size of each count is the product of the library size and the relative abundance of that gene in that sample. As input for data analysis, we expect count data in the form of a matrix of integer values. The matrix in the g-th row and the j-th column of the table tells how many reads have been mapped to gene g in sample j.
</sub>

Code for importing the data and defining the conditions (as last year tutorial).
Code for tximport when it is released.



## 4. DATA EXPLORATION AND QUALITY ASSESSMENT (ex3)

<sub>
Data quality assessment (QA) and exploration are essential steps of any data analysis. These steps should typically be performed very early in the analysis of a new data set, preceding or in parallel to the normalization step and differential expression testing. 
Our purpose is the detection of differentially expressed genes, and, in particular, we are looking for samples whose experimental treatment suffered from an abnormality that renders the data points obtained from these particular samples detrimental to our purpose. 

The QA processes run on the raw data might reveal technical issues, but other biases might still be present in your data and the best (only?) way to control for those is visual. For example, one could compare replicates (in which case, biological replicates are best), by: 
</sub>

-	looking differences in count distribution between samples, 
-	plotting a scatterplot of replicates, 
-	sample clustering

## 4.1. Data transformation

For data exploration and visualization, it is useful to work with transformed versions of the count data. As the count values distribution is highly skewed, the log2 transformation helps to approximately normalize the distributions.

Figure – Histogram of raw counts from 1 control sample (count on y-axis)

Log base 2 is typically used as it facilitates the conversion back to the original scale: a difference of 1 on the log base 2 scale corresponds to a fold change of 2 on the original count scale. Since count values for a gene can be zero in some conditions (and non-zero in others), we advocates the use of pseudocounts, i.e. transformations of the form
 
Y = log2(K+1); more generally Y = log2(K + k0), where K represents the count values and k0 is a positive constant.
 
Figure – Histogram of pseudocounts from 1 control sample (log2(count+1) on y-axis)

Even more, common statistical methods for exploratory analysis of multidimensional data, especially methods for clustering and ordination (e. g., principal-component analysis and the like), work best for (at least approximately) homoskedastic data; this means that the variance of an observable quantity (i.e., here, the expression strength of a gene) does not depend on the mean. 

In RNA-Seq data, however, variance grows with the mean, with larger variances for larger counts. For example, if one performs PCA directly on a matrix of “normalized” read counts, the result typically depends only on the few most strongly expressed genes because they show the largest absolute differences between samples. A simple strategy to avoid this is to take the logarithm of the “normalized” count values plus a small constant; however, now the genes with low counts tend to dominate the results because, due to the strong “Poisson” noise inherent to small count values, they show the strongest relative differences between samples.


In order to make counts approximately homoskedastic, the packages edgeR offers function to transform the data:

`cpm(..., prior.count = 2, log = TRUE){edgeR}`

WARNING: Note that these transformations are provided for applications other than differential testing. For differential testing to operate on raw counts and to use discrete distributions.


### 4.1.1. Boxplots per samples
The boxplot method provides an easy way to visualize the distribution of pseoudocounts in each sample. In the boxplot display, a box is formed with sides at the 25-th and 75-th percentiles of the distribution. A line is also drawn within the box at the level of the median. Whiskers are also drawn extending beyond each end of the box with points beyond the whiskers typically indicating outliers.

` Code Figure – Boxplots per sample (y-axis -> log2(count+1))`


### 4.1.2. MA-plot between samples
An MA-plot is a plot of log-fold change (M-values, i.e. the log of the ratio of level counts for each gene between two samples) against the log-average (A-values, i.e. the average level counts for each gene across the two samples). 
The MA-plot is a useful to visualize reproducibility between samples of an experiment. From a MA-plot one can see if normalization is needed. In MA plot, genes with similar expression levels in two samples will appear around the horizontal line y = 0. A lowess fit (in red) is plotted underlying a possible trend in the bias related to the mean expression. Below we show the code to produce a simple MA-plot (e.g. between control 1 and control 2 samples).

`## x – pseudocounts of sample1 , y – pseudocounts of sample2`
`## M-values M = x - y ## A-values A = (x + y)/2`
`Figure - plot M vs A`


### 4.1.3. Clustering of the sample-to-sample distances
To explore the similarities and dissimilarities between samples, it is often instructive to look a clustering heatmap of sample-to-sample distance matrix. 
A heatmap is a two-dimensional, rectangular, coloured grid. It displays data that themselves come in the form of a square matrix:

- the colour of each rectangle is determined by the value of the corresponding entry in the matrix, 
-	the rows and columns of the matrix are rearranged independently according to some hierarchical clustering method (Euclidean distance is the most used one), so that similar rows and columns are placed next to each other, respectively

`# CODE`
`Figure – heatmap of sample distances`


### 4.1.4 MDS of samples
The purpose of multidimensional scaling (MDS) is to provide a visual representation of the pattern of proximities (i.e., similarities or distances) among a set of objects. MDS takes a set of similarities and returns a set of points such that the distances between the points are approximately equal to the similarities. The similarities between each pair of samples (columns) is the root-mean-square deviation (Euclidean distance) for the top genes. Distances on the plot can be interpreted as leading log2-fold-change, meaning the typical (root-mean-square) log2-fold-change between the samples for the genes that distinguish those samples. The function plotMDS has an argument gene.selection which, set to "common", chooses as top genes those with the largest root-mean-square deviations between samples.

The plotMDS is a simple helper function from edgeR that performs the MDS and makes the plot

`## CODE of plotMDS on our samples`


***

made by Andrea and Kamil
