---
output: pdf_document
---
#Unix cluster computing

Last time we got familiar with basic concept of UNIX shell bash. Now we would like to learn how to compute on cluster and how to behave there. Here are more materials about the cluster we will use [Vital-it](http://edu.isb-sib.ch/course/view.php?id=72).

### OS X, Linux users

**connection**

Open terminal and connect to Vital-it using `ssh` (secure shell)
```bash
ssh <username>@prd.vital-it.ch
```

**file transfer**

To copy file **to** cluster you have to be in the terminal of your computer, then
```bash
scp <path>/<filename> <username>@prd.vital-it.ch:/<path>/<filename> # (scp stands for secure copy)
```

to copy file **from** the cluster to your computer, you still have to be in the terminal of your computer, but the order of locations in the command will be just reverse
```bash
scp <username>@prd.vital-it.ch:/<path>/<filename> <path>/<filename>
```

### Windows users

**connection**

1. Install [WinSCP](http://winscp.net/eng/download.php) and [Putty](http://www.chiark.greenend.org.uk/~sgtatham/putty/download.html)
2. First start WinSCP by:
    -	hostname: prd.vital-it.ch
    -	username: 
    -	password:
3. Then start your Putty entering hostname prd.vital-it.ch

**file transfer**

Transfer the file to your computer by drag and drop from the server side. Every time in the exercise you see `scp` command, it is only for Mac/Linux users, you use WinSCP. In the WinSCP navigate your paths and check your folders. 

***

\newpage

## Scripts

One of the advantages of the command line over programs based on mouse and clicking is, that is very easy to store, what exactly you have done in a simple text file. This file can be than executed again to achieve exactly same computation, what guarantees reproducibility. Those files are called scripts they usually have extension `.sh`. Writing commands you use in scripts also helps you to reuse it.

## Variables

`bash` also allows to use variables, similarly to `R` or `python`. You can define them by `<varible_name>=<value>`, where `<varible_name>` is usually by convention upper-case and `<value>` can be either a number or a string (text).

0. connect to Vital-it
1. define variable `VAR` to `1`

since terminal is primarily trying to execute commands, variables has to be called using a special symbol `$`. A `bash` command that simply prints its arguments is called `echo` (similar to `print` or `message` in `R`).

2. print your variable on terminal

```bash
echo $VAR
```

there are several pre-defined variables in the terminal defining a behaviour of the terminal called environmental variables. One of them is especially important. The variable `PATH` lists all locations of programs (binaries) that bash will search every time you type a command. You can check it using `echo` (do not forget `$` symbol)

3. look at locations of binaries

***

\newpage

## Pimp my bash

Your terminal is now white and when you type `ls` it is hard to get oriented. One wonderful feature how to make life in terminal easier is to colour it! You can do so by setting an environmental variable `PS1`.

1. Change colour of your terminal

```bash
export PS1='${debian_chroot:+($debian_chroot)}\[\033[01;32m\]\u@\h\[\033[00m\]:\
\[\033[01;34m\]\w\[\033[00m\]\$ '
```

<sub> You can write the command on one line only. The last symbol in line `\` means, that command continues on the next line, however the second `\` (first character of the second line) is needed for the expression to work properly. Environmental variable `PS1` defines how every new line will look like. For instance, if you specify `PS1=''`, prompt will disappear completely.
<sub\>

Great, terminal is colourful now, however, if you exit and start a new terminal session, it will be white again (this behaviour helps you not to mess with your own terminal). The way, how to make changes in terminal permanent is to edit a bash script `~/.bashrc`, which is executed every time you open new terminal window.

2. Add the line in the end of your `~/.bashrc` file

```bash
echo "export PS1='${debian_chroot:+($debian_chroot)}\[\033[01;32m\]\u@\h\[\033[00m\]:\
\[\033[01;34m\]\w\[\033[00m\]\$ '" >> ~/.bashrc
```

<sub>
Note, how do we add the line to the end of the `.bashrc`. Quotes `""` serve to tell bash, that he should not take the text inside as a command, but treat it as a plain text, echo just tells bash to print what is quotes and finally using `>>` operator we redirected the output stream to file `~/.bashrc`. `echo` automatically adds a newline character in the end of every message its printing. The very same effect could be achieved, if you open the document in a text editor (line `vi`) and copy-paste the line.
<sub\>

You can easily check how `~/.bashrc` works by adding there a line `echo I love sequence a genome class`. Every time, you will log to Vital-it you will see this important message. The dot at the begging of the file name means, that it is hidden file, therefore if you go to your home `cd ~` and list your files there `ls` you wont see it. You have to use `ls -a` option, which stands for all. In file starting by `.` is usually stored a local user setting of some program, meta files or history (like `.R` settings of R, `.bash_history` bash history, `.ssh` setting of SSH, etc.). In fact, we would like to know about hidden files almost all the time, but specifying an argument `-a` of `ls` every single time is just... annoying. The way out is a creation of aliases (shortcuts) to commands you frequently use. 

1. Create an alias by adding following line `alias ll='ls -lah'` to the end of `~/.bashrc`.

You can use from now on a command `ll` instead of `ls -lah` (list all, long format, in human readable units).

***

\newpage

## Prepare a workspace

Before anything, keep in mind that you can save a lot of energy and mistakes by using `<tab>` for auto completion of all existing paths you type and by using your command history (up and down arrows and `Ctrl+r`)

1. go to your scratch directory. If it does not exist for any reason, make the directory using `mkdir <dir_name>` (make directory). Check the content. What do you expect?

```bash
mkdir /scratch/beegfs/monthly/mls_2016/<username> #if it doesnt exist
cd /scratch/beegfs/monthly/mls_2016/<username>
ls -l
```

it is probably empty. `scratch/monthly/` is a working place, all files and folders older than one month are automatically deleted (all data from the last semester). From now on, keep all your big files here (or at your computer). The quota for disc `beegfs` is 5TB per user. The quota for `home` is 2GB. Therefore, all jobs run from working directories located at `/scratch/beegfs/monthly/mls_2015/<username>`.

2. **Copy your data to your working dir.**
3. **Get reference**

***

## Explore genomic reads

The files you got is a compressed textfile, if you try `head <file_name>`, you will get just meaningless characters. There are two solutions, either you can uncompress your files by command `gunzip` or preferably you can use command `zcat` to print the file on the standard output. Do not forget that it is still a half GB file each - a lot of letters, you better pipe them somewhere.

1. Format `.fastq` is just formatted plain text formats, so you could use all line tools for parsing text. Check the format with `head`

```bash
zcat <read_file.fastq.gz> | head
```

Every read is represented by four lines: header line, sequence line, quality of basecalling (in PHRED score) and comments (usually blank line).

2. How many reads are in the files `R1` and `R2`? Is it a same number?

***

\newpage

## Basic cluster operations

### Modules

1. Programs available on the Vital-it you can loaded using 

```bash
module add <field>/<category>/<program>/<version>
```

2. The list of programs and commands for loading them can be found on [Vital-it webpage](http://www.vital-it.ch/services/software). Find there a command for program fastqc and execute it.

<sub>Vital-it trick: An alternative option for webpage interface is a command `vit_soft`, `vit_soft -s <name>` will return a list of packages containing the <name> you specified. To get a command for loading package you have to choose a package from the list and execute `vit_soft -p <package_name>`<sub\>

3. Now, check how your `PATH` has changed. 

Adding module is nothing more than just adding a new path to list of directories searched for binaries.

### Running a simple job

1. You will need two main commands: `bsub` to submit a job and `bjobs` to obtain job status. Submission of a job have the following syntax 

```bash
bsub <bsub_parameters> <command> <command_parameters>
```

Vital-it is a computing cluster - hundreds of computers connected in a huge grid. The computer you are logging to, when you want to submit a job (prd.vital-it.ch) is one computer (server) and it serves exclusively for distribution of computational tasks across the computing grid. This computer is called front-end. Every time, you run a `bsub` on the front-end, it will put your job in the queue and assign you a computer once, there will be one free for you, given specifications you have provided in `<bsub_parameters>`. 

`fastqc` computes a quality control and outputs it to existing output folder defined by parameter `-o`. 

```bash
fastqc -o <fastqc_results> <sequencing_reads>
```

2. Create a folder `<fastqc_results>` for an output of `fastqc` and submit a job computing a quality control of read file you have copied.
3. Check the status of the submitted job

**BREAK**

4. Once it finishes, verify the existence of files. Copy those files to your computer (explained at beginning) and check the html file.

***

Credits: based on Nadja's material, made by Andrea and Kamil;
Contact: kamiljaron@gmail.com
