---
output: pdf_document
---

#Unix introduction

Unix is an essential tool for work with big amounts of data, this tutorial is trying to show the practical usage of the UNIX command line `bash` with emphasis on explanation of mechanisms and logic behind. More materials about UNIX and `bash` in particular you can find all over the internet, one I like is [Advanced Bash-Scripting Guide](http://tldp.org/LDP/abs/html/index.html).

## Syntax of this exercise

* The comment symbol of bash is `#`, everything after this symbol is ignored.
* Blocks of code will be in boxes, note also the usual syntax of bash commands `command -parameters location`.

```bash
ls -h / # list files and directories in -human readable format located at root directory
```

* Anything in `< >` you have to replace by meaningful paths, file names or parameters.

```bash
ssh <username>@prd.vital-it.ch #this will be written in this document
ssh kjaron@prd.vital-it.ch #this is example, how it can be used
```

* Bash commands are (usually) meaningful acronyms, the explanation is written after every first occurrence in brackets.
* Detail explanations will be written in blocks as follows

<sub>These explanatory blocks should help to understand. So, every time you feel confused, read this blocks. If it is not going to help, let me know.</sub>

## Getting started

Unix is a type of operating system, to communicate with the operating system we will use a command line language called `bash`.
Just like natural languages, computer languages have their own syntax and vocabulary.
The goal of this class is to provide you a good staring point to this important skill for modern science.

The very first step is to open a terminal.

***

\newpage

## Open terminal and get to cluster

Note that if you use Linux, Mac OS X or Windows 10 you have `bash` on your computer as well. However, the whole semester we will use the computational cluster for the most of tasks done within this class, therefore getting familiar with the computation cluster (Vital-it) is important too.

### Windows users

1. Install [WinSCP](http://winscp.net/eng/download.php) and [Putty](http://www.chiark.greenend.org.uk/~sgtatham/putty/download.html)
2. First start WinSCP by:
    -	hostname: prd.vital-it.ch
    -	username:
    -	password:
3. Then start your Putty entering hostname prd.vital-it.ch

<sub>
The programs mentioned above will allow you to log to cluster.
If you would like to have standard unix tools on your computer,
you can allow the beta version of Linux subsystem on Windows 10 (see [this tutorial](https://www.howtogeek.com/249966/how-to-install-and-use-the-linux-bash-shell-on-windows-10/)) or you can install [cygwin](http://www.cygwin.com/), a collection of unix tools adopted for windows.
</sub>

### Linux / OS X users

Open a terminal. Pressing `ctrl + alt + t` should do the job on common Linux systems.
On Max OS X you can press `cmd + space` to start a spotlight search,
then type `terminal`.
This should find the terminal app therefore you can just hit `enter` to launch it.

Now, connect to Vital-it using `ssh` (secure shell).

```bash
ssh <username>@prd.vital-it.ch
```

You will be asked to type a password.

***

\newpage

## Bash as explorer

You should see a command line in front of you.
Now you should to understand how to use `bash` to manipulate files just like in an explorer or finder.
It is good to mention that there is no trash in `bash`, if you rewrite or remove any file, it's practically lost.
You do not have to be scared to explore on Vital-it,
it is not a problem to restore your cluster account to the default setting anytime.

### Browsing directories

1. Go to your home directory `/home/<user_name>` and check the content using `ls` (list segments). You will probably see nothing (i.e. your home is empty).
2. Make a new directory, enter the directory

```bash
mkdir SAGE
cd Sage
```

<sub>
Everything in UNIX is case sensitive, i.e. the name Sage is not the same as SAGE. Therefore the error message you see `cd: Sage: No such file or directory`. Error messages are usually very precise in what is wrong and the vast majority of problems is caused by execution of a command in wrong directory.
</sub>

3. Use `pwd` (print working directory) to check your location.
4. Correct the command `cd Sage` to actually enter the directory `SAGE`.
5. Use `pwd` to check your location.
6. Go to parent directory `cd ..` (change directory) and verify the success using `pwd` command again. Generally, `cd` changes a working directory to the provided path `cd <path>`.

<sub>The command `cd` (change directory) and correct handling of paths is essential for everything done in Unix terminal. The path can be defined relatively to your current directory (ex: `cd ./SAGE` or `cd ..`) or absolutely, i.e. starting from the root directory of the computer (ex: `cd /home/kjaron`). Meanings of symbols are as follows: `.` is a current directory, `..` is a parent directory, `/` is a root directory, `~` is your home directory. Note that symbol `~` carry different path for every user: `/home/<username>` and symbols `.` and `..` carry different path in every single directory, hence a path specified using one of these two symbols is called relative path. </sub>

### Download, copy, move, rename and remove

1.Â Copy directory `/scratch/beegfs/monthly/mls_2017/UNIX_testfiles` to your working directory using command cp (copy)

```bash
cp -<parameters> <path/file> <path/new_file_name> # generally
cp -r /scratch/beegfs/monthly/mls_2017/UNIX_testfiles .
```

<sub> If you want to copy to retain the original name, you can specify just the path.
If you want to copy a file (or directory) to the working directory you are in,
you can just write only `.`.
You should be also aware of why you had to specify `-r` to copy a directory </sub>

3. Copy the file to your working directory using command `cp` (copy)
4. Go inside and check content of the directory (you should know how, if you don't, scroll up).
5. Copy file `browsing.txt` in the directory `nonemptydir`. Do you need the parameter `-r`? If you are not sure, check manual page (`man cp`, for all inbuilt commands there is a manual page you can check).
6. Go to `nonemptydir`, rename `browsing.txt` to `<your_new_name>.txt` and move the renamed file to `UNIX_testfiles`.

```bash
cd ./nonemptydir
mv browsing.txt <your_new_name>.txt
mv <your_new_name>.txt ..
cd ..
```

6. What is the difference between second and the third line? Can you figure out how to do the same without changing the directory to `nonemptydir`?

<sub>Command `mv` (move) moves file to the given path and renames it to the given name. So, if the `<path_to_input>` is the same as `<path_to_output>` in the command

```bash
mv <path_to_input>/<file_name> <path_to_output>/<new_file_name>
```

, the file will be only renamed. If the path is different, but `<new_file_name>` is not provided, the file will be just moved (not renamed).</sub>

7. Remove the copied and renamed file.

```bash
rm <your_new_name>.txt
```

8. Try to remove directory `nonemptydir`. Could you figure out why `rm <dir>` and `rmdir <dir>` is not working?

<sub> Command `rm` is to remove files, `rmdir` to remove empty directories. If you want to remove a directory with files inside, you have to specify it `rm -r <dir_to_delete>`. However, be careful, there is nothing like "Trash" in the bash, i.e. what is deleted is permanently lost! </sub>

9. Try command `man <command>` (manual) to show manual page of some commands. What does the `-a` mean in the `ls` command? Could you use this parameter to find out why `rmdir` is not working on `nonemptydir`, even through `ls` is not showing any files inside?

***

## Break!!!

\newpage

## More than explrorer

There are two types of commands you can use in `bash`. Inbuilt and default commands (usually with `man` page). And programs you have installed. "installed" on Linux means usually that a binary file is copied to location where `bash` searches for commands. The locations of commands (programs) can be shown by typing command `echo $PATH`. However, before going to explore it, we will get familiar with standard commands of bash.

### Commands text processing

There are many many ways how to show a plain text file in the terminal, so if you find another solutions than we provide, do not wonder. Some of commands for text processing:

- Commands `head -n <num_of_lines> <text_file>`, `tail -n <num_of_lines> <text_file>` and print the beginning (first n lines), the end of input files (last n lines), respectively.
- Command `less <text_file>` will allow to browse text file (of any size). You can use scroll or use arrows to browse the file. If you press `q`, you will escape back to the terminal. If you want to explore, `less` is very good tool.
- Command `wc -lwc <text_file>` (word count) is counting number of newline characters, words and characters in the input file. You can specify output by parameters .
- Command `tr -d '<string_to_be_removed>' <text_file>` (translate characters) can delete all occurrences of a given string (Note: a character is also a string).

1. Use commands `head -n <num_of_lines> <text_file>`, `tail -n <num_of_lines> <text_file>` and `cat <text_file>` (catenate) to see the beginning, the end or whole file `browsing.txt`, respectively. How many lines are in the file?
2. Use `less <text_file>` to check `AllIllumina-PEadapters.fa`. You can use scroll or arrows to move in the file. If you press `q`, you will escape back to terminal. You can also use `head` and `tail` commands to see the beginning and the end of the file.
3. To get an idea, about the file, you can use command `wc` (word count), which is counting number of, newline characters, words and characters in the input file. You can specify output by parameters. How many lines are in `browsing.txt`?

Command `grep` searches the file and prints all lines containing a pattern specified by a parameter: `grep <pattern> <file.txt>`. I heavily recommend to write such patterns in quotes "", because some characters have a special meaning in the command line (for example white spaces, or the symbols ">", "|", ".", "\\", etc.).

4. Explore behaviour of `grep` selecting different lines in file `browsing.txt`. Example:

```bash
grep "to check" browsing.txt
```

One of the most powerful and simple tools for reading text files in terminal is `less`. It is loading to memory only parts of opened file that are read, therefore you can use this tool to check a file of any size (even files exceeding RAM of your computer). To browse file you can use arrows, page-up/down or scroll of a mouse.

5. Look at the `sample.fasta` file using program `less <file_to_read>`

### Pipes

The output of commands is usually printed to terminal (list of files in case of `ls`, content of file in case of `cat`, etc.). Using symbol `|` you can redirect ("pipe") the standard output as an input to another command. We can use a few commands to find out how many Illumina adapters are in the file `AllIllumina-PEadapters.fa`. File `AllIllumina-PEadapters.fa` is a fasta file, so every sequence is represented by header line starting by symbol ">" and the sequence itself.

1. How many adapters are saved in the file `AllIllumina-PEadapters.fa`? [22]

```bash
grep ">" AllIllumina-PEadapters.fa | wc -l
```

To redirect an output to a file rather than to other command, you can use the symbol `>`. Be careful, if the target file exists, it will be overwritten. To add lines to the end of the file, you can use `>>` (= concatenate).

Using the same tools we answer various questions. Another example of usage to compute a number of nucleotides in simple fasta.

2. How many nucleotides are in file `sample.fasta`? [66058] Hint: check `-v` argument of grep.

<sub>
If you ended up on number 66719 you have counted a newline characters as nucleotides. You can either count them by `wc -l` or you can use a trick described at task 4.
</sub>

3. The very same problem you can also solve using command `tail` which can also print all but a given number of skipped lines from the beginning `tail -n +<number_of_the_first_line_to_be_printed> <file>`. Following command prints number of characters.

```bash
tail -n +2 sample.fasta | wc -c
```

However, this command shows number of characters (newline characters included). It is not entirely correct.

4. We can use `tr` (translate characters) command to delete all newline characters before you count them using `wc`. Note how the commands are nested, you can use as many pipes as you need.

```bash
tail -n +2 sample.fasta | tr -d '\n' | wc -c
```

<sub>
There are many ways how to read or edit plain text files using command line tools. Next to the presented commands, there are several scripting languages specialized for text processing like `awk` or `sed` or `grep`. In bioinformatics is also common to use `biopython` or `BioPerl` to work with fasta and fastq files.
</sub>

## Challenges

1. Extract all header lines of `AllIllumina-PEadapters.fa` file to separated file. If you do the same on `sample.fasta`, how many headers you extract?

```{bash echo=FALSE, eval=FALSE}
grep ">" AllIllumina-PEadapters.fa > adapter_headers.txt
wc -l adapter_headers.txt
```

2. Compute total length of low-complexity nucleotides (lower-case letters, i.e. non-capital letters) in `sample.fasta` [33576 I think].

```{bash echo=FALSE, eval=FALSE}
grep -v ">" sample.fasta | sed 's/[A-Z]//g' | tr -d '\n' | wc -c
```

3. Command `touch` will modify a time of last modification of a file. Find a way, how to touch all your files, verify the success using `ls -l`. Hint : Use google to find an answer and manual to understand it.

```{bash echo=FALSE, eval=FALSE}
find . -exec touch {} \;
```

<sub>There are loads of different solutions. One of them can be found in [the source code for this document online](https://github.com/KamilSJaron/Sequence-a-genome/blob/master/2017_autumn/1_unix_I/1_unix_I.Rmd) <sub\>

***

## Tips and tricks

- Command `cd` without any argument changes the working directory to your home. Symbol `~` works the same, it is symbol of home.
- Press tabulator to call whisperer (auto completion) every time you are writing a path. It is handy with long file names as we can find in genomics. Note that you will have to type all the first letters of the file which are in common with other file names of same the directory before auto completion can be achieved by pressing tabulator. It is a good way how to prevent typos, because auto completed path is usually valid.
- Arrows up and down allows to browse your command history saved at `~/.bash_history`
- Press `Ctrl+R` for a full text search in the command history
- Press `Ctrl+C` to stop the running process, useful if you use `cat` on a huge file to stop seeing text flowing endlessly... Beware to remove result/temporary files created by the program if you use `Ctrl+C` to stop a running program, otherwise there may be conflicts the next time you run the same program.
- You can use wild characters `?`, `[]` and `*`, to specify a group of files. `?` stands for any single character, `[XYZ]` one of the specifies characters (XYZ in this case, you can also use numbers), `*` any number of any characters. See examples

```bash
cat *.txt # prints all text files
ls ?? # lists all files or directories with exactly two letter names
cp /<path>/LS[1234]* . # copiing all files staring by LS1 LS2 LS3 or LS4 from given <path>
```

***

Credits: made by Kamil, based on Andrea's and Nadja's materials from previous years;
Contact: kamiljaron@gmail.com
