#RNA seq

Today, we will check the quality of the sequencing reads. If it is fine, we will build a reference and finally we will map all reads to the reference to get expression levels.

## Last time, in Sequence a genome...

- remote connection and file transfer `ssh`, `scp` to server `<username>@prd.vital-it.ch`
- basic commands `pwd`, `ls`, `cd`, `cp`, `mv`, `mkdir` and finally `man` if you forgot anything
- text processing tools `cat`, `head`, `tail`, `less` and `vi`
- operations on cluster `module add`, `bsub` and `bjobs`
- `fastqc`, `kallisto` and `R` prepared to use (on vital-it or your computer)

***

## Vital-it

Useful commands and options:

- bsub: submit a job on Vital-IT
    *   -J: job name
    *   -n: number of cores to use for the program
    *   -R: amount of space and number of cores to reserve for your job
- bjobs: see your running jobs
- bkill: kill the job

## Quality control

0. Log to Vital-it
1. Go to your working folder and create output directory `fastqc_results`
2. Run fastqc. You can put several fastq files with empty space between them.

```
bsub –R “rusage[mem=6000] span[ptile=2]” –n 2 –J <job_name> fastqc –t 2 <fastq_file> -o <output_folder>
```

3. Copy fastqc results to your computer
4. Open the `fastqc_report.html` and check and interpret the different graphs. How many reads do you have? Are the reads single or paired-end? Do all the reads have a good quality?

***

## Bulding a reference

***

## Starting a mapping

***

made by Andrea and Kamil
