#RNA seq

Today, we will check the quality of the sequencing reads. If it is fine, we will build a reference and finally we will map all reads to the reference to get expression levels.

## Last time, in Sequence a genome...

### unix reminder

- remote connection and file transfer `ssh`, `scp` to server `<username>@prd.vital-it.ch`
- basic commands `pwd`, `ls`, `cd`, `cp`, `mv`, `mkdir` and finally `man` if you forgot anything
- text processing tools `cat`, `head`, `tail`, `less` and `vi`
- operations on cluster `module add`, `bsub` and `bjobs`

### data exploration

- genome, annotation and extraction of coding regions
- `fastqc`, `kallisto` and `R` prepared to use (on vital-it or your computer)
- quality control of reads using `fastqc`

***

## Indexing the reference

(for all groups same)

1. Go to your working directory (`/scratch/montly/...`). Make a folder (`mkdir`) for `kallisto` and go there (`cd`)
2. `kallisto` has a mode for an indexing called `index`. Parameter `-i` specifies an output name for the indexed file. Use `kallisto` to index the file with coding regions.

```bash
gsub kallisto index -i <indexed_cds.idx> ./<S5_cds.fasta>
```

<sub>
<span style="color:red">some explanation of indexing??</span>
</sub>

***
3. check if you have a `.idx` file present in directory.

## Starting a mapping

How was the library prep going? What was the average size of fragments in libraries? Its standard deviation? These informations are used by kallisto...

```{r set-options, echo=FALSE, cache=FALSE}

out <- data.frame(group = 1:16, condition = rep(c('LM','SA','WL','WR'), each = 4), bio_repl = rep(c(1:4),4))
out$path_to_reads <- paste('/scratch/cluster/monthly/kjaron/sag/RNAseq',out$con, paste(out$con,out$bio_repl,'/', sep = ''), sep = '/')
out$mean <- 350
out$sd <- 60
suppressWarnings(library(knitr))
kable(out)
```

Now, we want to map the RNAseq reads of each biological replicate to indexed coding regions, we also want to save pseudoalignment in the `.bam` format for visualisation (IGV). You have to specify: `<mode>` to be `quant`, `-i` indexed file with coding regions, `-o` output folder (will be created if does not exist), `-b` number of bootstraps, the type of reads `--single` ( = single end reads), `-l` mean fragment size in library, `-s` standard deviation of fragment sizes in library, `--pseudobam` is parameter specifiing that a pseudoaligment should be created and printed on standard output and finally all input files. 

```bash
kallisto <mode> -i  <indexed_cds.idx> -o <output_folder> -b <number_of_bootstraps> --single -l <mean> -s <sd> --pseudobam <input_files>
```

Since we want to capture a `.bam` file and we want it ordered, we pipe it to `samtools`.

```bash
kallisto ... | samtools view -bS - | samtools sort - -o <output.bam>
```

All reads are still saved at `/scratch/cluster/monthly/kjaron/sag/RNAseq/<condition>/<biorep>/`. You can again use wild character `*` to select all gzipped fasta file in the folder `*.fasta.gz`. So, putting all together gives us

```bash
gsub kallisto quant -i  <indexed_cds.idx> -o <output_folder> -b 100 --single -l <mean> -s <sd> --pseudobam <input_files> | samtools view -bS - | samtools sort - -o <output.bam>
```

This will take several minutes.

***

<span style="color:red">Formating bam file for IGV. </span>

```bash
samtools index <output.bam>
```

***

IGV time - copy a pseudoalignment file (`.bam`), the genome (`.fasta`) and the annotation (`.XXX`) to your computer. Start IGV.

<span style="color:red">IGV instructions (Andrea's pictures)</span>

made by Andrea and Kamil
