#RNA seq

<sub>
Motivation RNA sequencing (RNA-seq) technologies have recently come to prominence as technique for characterizing RNA transcripts and comparative analyses of their abundances. It is used in many areas of biology including functional genomics, developmental biology and cancer biology. In RNA-seq, mapped reads are typically aggregated to counts at some level of interest, such as transcripts, exons, or genes. The count for a given gene or region quantifies the expression of that gene or region. We are very often interested in which genes or regions have different levels of expression under different experimental conditions – this is known as differential expression. The nature of the count data produced in RNA-seq experiments introduces challenges for the analysis of differential expression. The computational and statistical methods for analyzing RNA-seq data are recent, and successful methods for the analysis of microarray data are not directly applicable to RNA-seq data. It seems that dedicated tools, designed especially for count data, will be required to extract the best possible information from RNA-seq data. Several emerging tools within Bioconductor (a repository for open-source software for bioinformatics) have been developed for the differential analysis of count data, including DESeq, DESeq2, edgeR, baySeq, limma (voom function), …but best practices are not yet clearly established.
</sub>

Today, we will check the quality of the sequencing reads. If it is fine, we will build a reference and finally we will map all reads to the reference to get expression levels.

## Last time, in Sequence a genome...

- remote connection and file transfer `ssh`, `scp` to server `<username>@prd.vital-it.ch`
- basic commands `pwd`, `ls`, `cd`, `cp`, `mv`, `mkdir` and finally `man` if you forgot anything
- text processing tools `cat`, `head`, `tail`, `less` and `vi`
- operations on cluster `module add`, `bsub` and `bjobs`
- `fastqc`, `kallisto` and `R` prepared to use (on vital-it or your computer)

***

## Vital-it

Useful commands and options:

- bsub: submit a job on Vital-IT
    *   -J: job name
    *   -n: number of cores to use for the program
    *   -R: amount of space and number of cores to reserve for your job
- bjobs: see your running jobs
- bkill: kill the job

## Quality control

0. Log to Vital-it
1. Go to your working folder and create output directory `fastqc_results`
2. Run fastqc. You can put several fastq files with empty space between them.

```
bsub –R “rusage[mem=6000] span[ptile=2]” –n 2 –J <job_name> fastqc –t 2 <fastq_file> -o
```

3. Copy fastqc results to your computer
4. Open the `fastqc_report.html` and check and interpret the different graphs. How many reads do you have? Are the reads single or paired-end? Do all the reads have a good quality?

***

## Bulding a reference

***

## Starting a mapping

***

made by Andrea and Kamil
