---
title: "Sequence a genome"
author: "Kamil S JaroÅˆ"
date: "Spring 2016"
output: ioslides_presentation
---

## What have we done last semester?

  - isolated 2 strains of Pseudomonas
  - sequenced both genomes
  - annotated genes of interest

## What have we done last semester?

  - isolated 2 strains of Pseudomonas
  - sequenced both genomes - almost identical genomes
  - annotated genes of interest - genebank file

## What have we done last semester?

  - isolated 2 strains of Pseudomonas
  - sequenced both genomes - `.fasta` file
  - annotated genes of interest - `.gbk` file

## This semester

>- 24 . - 26.02.  : RNA sequencing (done)
>- 12.04. : **Data exploration**
>- 19.04. : Mapping reads and quantifying
>- 21.04. : Downstream analyses
>- 26.04. : Downstream analyses
>- 03.05. : Analysis of annotated genes
>- 10.05. : Analysis of annotated genes
>- 17.05. : Presentations

## Today

- 12.04. : Data exploration

    >- exploration of the data
    >- overview of experiment
    >- installation of all needed programs
    >- formation of working groups
    >- quality control

## Unix

is not your enemy!

Files on switch (myunil > classes > sequence a genome II):

- The manual for this practicals: `RNAseq1_unix.html`

- <span style="color:red">The unix basics exercise:</span> `unix_refresh.html`

- This presentation: `introduction.html`


## Syntax of this exercise

* The comment symbol of Bash is `#`, everything after this symbol is ignored
* blocks of code (bash or R) will be in boxes, note also the usual syntax of Unix commands `command -parameters location`.

```bash
ls -h / # list files and directories 
# in -human readable format 
# located at root directory
```

## Syntax of this exercise

* Anything in `< >` you have to replace by meaningful paths, file names or parameters.

```bash
ssh <username>@prd.vital-it.ch #this will be written in this document
ssh kjaron@prd.vital-it.ch #this is example, how it can be used
```

## Syntax of this exercise

* Bash commands are meaningful acronyms, the explanation is written after every first occurrence in brackets.
* Detail explanations will be written in blocks as follows

<sub>These explanatory blocks should help you to get oriented in documents of Sequence a genome class and to give you an idea how to read it. So, every time you feel confused, please read this blocks and if it does not help, ask one of the assistants.</sub>

## Vital-it

A cluster for scientific computing shared between four universities.

![](cluster_storage.png)

## Vital-it
scratch

```
/scratch/cluster/monthly/ # any files older than month are erased
/scratch/cluster/weekly/  # any files older than week are erased
/scratch/cluster/daily/   # any files older than day are erased
```

No backup

1TB quota per user

## Vital-it

Let try to get there...

```bash
ssh <username>@prd.vital-it.ch
```

## Vital-it

Let try to get there...

```bash
ssh <username>@prd.vital-it.ch
```

Check your home
```bash
cd /home/<username>/
ls -la
```

## Vital-it

Let try to get there...

```bash
ssh <username>@prd.vital-it.ch
```

Check your scratch
```bash
cd /scratch/cluster/monthly/<username>
ls -la
```

## Vital-it

Let try to get there...

```bash
ssh <username>@prd.vital-it.ch
```

Check your scratch
```bash
cd /scratch/cluster/monthly/<username>
ls -la
```
Error? Make your scratch then
```bash
mkdir /scratch/cluster/monthly/<username>
```

## Vital-it

Long story short

`/home/<username>` - scripts, configuration files

`/scratch/cluster/monthly/<username>` - data files, analyses

## Your turn...

 command         meaning          command    meaning
---------    ----------------    ---------  ---------
`ls`            list files         `cp`       copy
`pwd`        print working path    `mv`       move
`cd`         change directory      `rm`      remove
`man`           manual             `wc`     word count

## Experimental design

![](FIGURE1.png)

## Experimental design

![](complex_experimental_design.png)

## What is done already this semester?

You have...

- isolated RNA under different conditions
- created libraries, sequenced those libraries

>- How many conditions?
>- How many biological replicates?

## Our experimental design

 condition      full name             batch
-----------   ---------------       ---------
   LM          liquid medium        lane 1
   SA              sand             lane 1
   WL           wheat leafs         lane 2
   WR          wheat roots          lane 2


4 biological replicates each

##{}

```{r set-options, echo=FALSE, cache=FALSE}

out <- data.frame(group = paste('#',1:8, sep = ''), 
                  condition = rep(c('LM','SA'), each = 4), 
                  bio_repl = rep(c(1:4),2), 
                  group_ = paste('#',9:16, sep = ''), 
                  condition_ = rep(c('WL','WR'), each = 4),
                  bio_repl_ = rep(c(1:4),2))
suppressWarnings(library(knitr))
kable(out)
```

## Quality control

the first computationaly intensive job we have.

## Vital-it

![](cluster_cp.png)

`prd.vital-it.ch` (the front-end of Vital-it):

- directly on front-end: small tasks
- submit to cluster (anything long or computational intensive)

## Quality control on Vital-it

to submit a job

```bash
bsub -<parameters> <command> -<command_parameters>
```

- Never sumbit job from your home.

- Do not allow to job to write to the home.

to get a status of all your active jobs
```bash
bjobs
```

## Extract coding regions 

while quality control of your reads is computed...

```{r fig.width=6, fig.height=4,echo=FALSE, fig.cap="Classical RNA-seq pipeline"}
library(png)
library(grid)
img <- readPNG("./classicalRNAseq.png")
 grid.raster(img)
```

## Extract coding regions 

...lets extract the coding regions for kallisto.

```{r fig.width=6, fig.height=4,echo=FALSE, fig.cap="Kallisto RNA-seq pipeline"}
library(png)
library(grid)
img <- readPNG("./kallistoRNAseq.png")
 grid.raster(img)
```

## Quality control evaluation

I will open one of the reports.

## Quality control evaluation

Is it ok, what we have got?

## Adapter sequences identified by fastqc

>SmallRNA3pAdapter_1.5\
ATCTCGTATGCCGTCTTCTGCTTG

>PCRPrimerIndex1\
CAAGCAGAAGACGGCATACGAGATCGTGATGTGACTGGAGTTC

>RNAPCRPrimerIndex47\
CAAGCAGAAGACGGCATACGAGATCTTCGAGTGACTGGAGTTCCTTGGCACCCGAGAATTCCA

>IlluminaSingleEndAdapter2\
CAAGCAGAAGACGGCATACGAGCTCTTCCGATCT

>PEPCRPrimer2.0\
CAAGCAGAAGACGGCATACGAGATCGGTCTCGGCATTCCTGCTGAACCGCTCTTCCGATCT
