---
output: html_document
---
#Unix

During this semester, we will perform analyses requiring a basic knowledge of bash terminal. If you do not feel, that you are familiar with it, you probably want to check `unix_refresh.html`. Still here? Good, read please following section about syntax which will be common in all sessions we will have this semester.


## Syntax of this exersize

* blocks of code (bash or R) will be in boxes, note also the usual syntax of Unix commands `command -parameters location`. Comments in code are after `#` symbol.

```bash
ls -h / # list files and directories in -human readable format located at root directory
```

* The comment symbol of Bash is `#`, everything after this symbol is ignored 
* Anything in `< >` you have to replace by meaningful paths, file names or parameters.

```bash
ssh <username>@prd.vital-it.ch #this will be written in this document
ssh kjaron@prd.vital-it.ch #this is example, how it can be used
```

* Bash commands are meaningful acronyms, the explanation is written after every first occurrence in brackets.
* Detail explanations will be written in blocks as follows

<sub>These explanatory blocks should help you to get oriented in documents of Sequence a genome class and to give you an idea how to read it. So, every time you feel confused, be read this blocks and if it does not help, ask one of the assistants.</sub>

***

## Check your wokring folder

1. connect to Vital-it
2. go to your home directory `/home/<user_name>` and check the content using `ls` (list segments).
3. go to your working directory. If it does not exist for any reason, make the directory using `mkdir <dir_name>` (make directory). Check the content. What do you expect?

```bash
mkdir /scratch/cluster/monthly/mls_2015/<username> #if it doesnt exist
cd /scratch/cluster/monthly/mls_2015/<username>
ls -l
```

it is probably empty. `scratch/monthly/` is a working place, all files and folders older than one month are automatically deleted (all data form the last semester).


***

## Get the genome, locate the reads

1. go to your working directory (`/scratch/cluster/monthly/mls_2015/<username>`).
2. copy directory `/scratch/cluster/monthly/kjaron/genome` to your working directory using command `cp` (copy)

```bash
cp -r /scratch/cluster/monthly/kjaron/sag/genome .
```

3. Go inside and check the content (`ls`). You should see genome in `.fasta` format, genome annotation in `.gdb` and `.gtf` formats.

All RNA-seq files, are saved in `/scratch/cluster/monthly/kjaron/sag/RNAseq/<condition>/<biological_replicate>.fastq.gz`. 

4. Go to the folder, check the content. How many conditions we have got? To have overview of sizes of all folders and the sum of sizes of their content, you can use `du -h`.
5. Exploring further, how many biological replicates per condition we have got? 

<sub>
Genome files are relatively small, therefore it is fine, if every one of you will have a copy of the files. However, the raw fastq reads have 34GB, so do not copy them to your folder, rather use full path `/scratch/cluster/monthly/kjaron/sag/RNAseq/...` to the files, we will share.
</sub>

***

## Explore those files

Some command line tools for text processing:

- Commands `head -n <num_of_lines> <text_file>`, `tail -n <num_of_lines> <text_file>` and print the beginning, the end of input files, respectively.
- Program `less <text_file>` will allow to browse text file (of any size). You can use scroll or arrows to browse the file. If you press `q`, you will escape back to terminal. If you want to explore, `less` is very good tool.
- Command `wc -cwl` (word count) is counting number of characters, words and lines in the input file. You can specify output by parameters.
- Command `tr` (translate characters) can (not only) delete all occurrences of string.

1. Formats `.fasta` and `.gds` are just formatted plain text formats, so you could use all mentioned command line tools for parsing text. Try it, what is inside of the `.gds`? 

2. How many nucleotides are approximately in the genome? File `Pseud_protegens_S5_final.fasta` is a simple fasta file, so number of nucleotides is equal to number of all no newline characters in the file without the first line (header). So, if we ignore the header of the fasta you can just use `wc -c <file>`.
3. Since, the most of bash command line tools are very simple, you can easily combine them. The output of commands is usually printed to terminal (list of files in case of `ls`, head of a file in case of `head`, etc.). Using symbol `|` you can redirect ("pipe") an output as the input to an another command. We can use a few commands to find out how many nucleotides are precisely in the genome of *Pseudomonas protegens* have sequenced last semester.

If the argument `-n` of command `tail` is in format `+<first_line_to_print>`, it prints all lines starting by the `<first_line_to_print>`th one, to the end of file. To ensure, that in the file are no newline characters `\n`, we can use `tr -d <file>`, in the end we will use the came command for computing the number of characters.

```bash
tail -n +2 <file> | tr -d '\n' | wc -c
```

So, how big the genome is? Do you remember approx the size of the *P. protegens* genome from the last semester? Is it consistent with your knowledge?

<sub>
The sequences in `.fasta` files are usually stored in 60 nucleotides per line, but bioinformaticians are messy and the format of the `.fasta` slightly differs sometimes. In the case of our genome there are no newline characters in the sequence (i.e. the whole genome is in the second line of the `.fasta` file), you can check this by deleting `| tr -d '\n'` from the command. Note, that almost all software are automatically wrapping too long line to make a document more readable. Using `less Pseud_protegens_S5_final.fasta` can give you a feeling of a usual `.fasta` even we already know, that is not the case. Also, if our assembly would have several contigs, there should be more header lines in the files, so you would have to use different trick to filter nucleotide sequences only..
</sub>

<sub>
There are many ways how to read or edit plain text files using command line tools. Next to the presented commands, there are several scripting languages specialised for text processing like `awk` or `sed`. In bioinformatics is also common to use `biopython` or `BioPerl` to work with `.fasta` and `.fastq` files.
</sub>

***

Now we know what we have got, so it is a good time to stop typing commands and to start to ask... What are we doing? What do we want to do? What is the meaning of life?

<span style="color:red">(if there are no formed groups already, form groups)</span>

***

## Extract cds

1. go to `<your_working_directory>/genome/`
2. activate Vital-it module `cufflinks`

```bash
module add UHTS/Assembler/cufflinks/2.2.1;
```

3. use `gffread` from cufflinks to extract coding regions from `.fasta` and `.gtf` files

```bash
gffread  -g <S5_genome.fasta> -x <S5_cds.fasta> <S5_annotation.gtf>
```

4. Check the file `<S5_cds.fasta>` using `less` or `head`. Does it look like extracted coding regions? 
5. How many transcripts we have?

```bash
awk '/>/{ print }' <S5_cds.fasta> | wc -l
```

<sub>
`awk` is scripting language for parsing text files. What is in curly brackets `'{ command }'` is executed on every line. If there is a pattern written in slashes before, the command will be executed only on lines with this pattern. In our case, the command print whole line (`print`) is executed on all lines containing symbol `>`, which represent lines with headers. Counting number of headers will give us a number of cds we extracted.
</sub>

**bonus** could you extract lengths of all reads? Hint: biopython provides a very simple solution.
```{python echo=FALSE, eval=FALSE}
from Bio import SeqIO 

for seq_record in SeqIO.parse(<cds.fasta>, "fasta"):
    print(len(seq_record))
```

***

# Install needed software

You will need `R` and `IGV` on your computer and programs `fastqc` and `Kallisto` prepared to use on Vital-it or on your computer (OS X, Linux). Make your mind if you want to run programs locally or on the cluster and follow one of the guides.

- `R` - is expected to be already installed on your computer.
- `IGV` - is literally Integrative Genomics Viewer. Follow instructions on the [webpage](https://www.broadinstitute.org/software/igv/download#binary) to install `IGV`.
- `fastqc` - is computing statistics of ngs reads for quality control, specifics bellow.
- `Kallisto` - is quantifying abundances of transcripts from RNA-seq data, specifics bellow.
- `samtools`


### OS X

`Kallisto` - A different ways how to get it can be found on [their webpage](https://pachterlab.github.io/kallisto/starting.html). We will install Kallisto using homebrew - package manager for OS X. Homebrew you can install using following command

```bash
ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
```

<sub> command `curl` downloads an installation script from the URL and ruby is a scripting language which will execute downloaded script. (Ruby is some ways similar with Pearl or Python). Analogical command to `curl` on Linux is `wget`.</sub>

than switch to scientific collection and install kallisto

```bash
brew tap homebrew/science
brew install kallisto 
```

`fastqc` is also included in scientific repository of homebrew

```bash
brew install fastqc
```

### Linux

`Kallisto` for Linux have to be build from source files, no worries, it sounds harder that it is. Use [building manual](https://pachterlab.github.io/kallisto/source.html) and if you get stack or if you have any question, do not hesitate, ask.
`fastqc` is written in Java, check if you have installed Java v1.6 or higher. 
```bash
java -version
```
if is it fine, download fastqc from [Babraham bioinformatics webpage](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/). Unzip the file. Go to the folder, change permissions to make `fastqc` executable and link it to `/usr/bin/`.
```bash
chmod u+x fastqc
sudo ln -s /path/to/FastQC/fastqc /usr/local/bin/fastqc
```

<sub>as you might know, if you are using Linux, `sudo` will execute command with administrator rights, therefore the command will ask you for a password to your account. Also, if your account has no administrator rights you wont be able to link the executable to your `/usr/bin/`. In this case you can add path to `fastqc` to your PATH variable as it is described in vital-it section or you can just run it using full path to executive.</sub>

### Vital-it

1. Log to vital-it. Open file `/home/<your_user_name>/.bashrc` in `vi`. 

`vi /home/<your_user_name>/.bashrc`

2. Type `a` to enter writing mode, then copy to the end of the script these lines:

```
module add UHTS/Analysis/kallisto/0.42.4;
module add UHTS/Quality_control/fastqc/0.11.2;
module add UHTS/Analysis/samtools/latest
```

3. Press `esc` to escape the writing mode, press `:` to enter command mode, type `wq` (write and quit) and confirm by pressing `enter`.
4. Activate changes by running

```bash
source /home/<your_user_name>/.bashrc
```

<sub> script `.bashrc` located at your home directory is a bash script which is executed every time you start a new session of terminal. You can easily check how it works by adding there a line `echo I love sequence a genome class`. Every time, you will log to Vital-it you will see this important message. The dot at the begging of the file name means, that it is hidden file, therefore if you go to your home `cd ~` and list your files there `ls` you wont see it. You have to use `ls -a` option, which stands for all. In file starting by `.` is usually stored a local user setting of some program, meta files or history (like `.R` settings of R, `bash_history` bash history, `.ssh` setting of SSH, etc.). </sub>

***

## Install needed R packages

### any OS

Start a R on your computer (NOT on vital-it) and install following packages from bioconductor:

* `edgeR` on [webpage](https://bioconductor.org/packages/release/bioc/html/edgeR.html)
* `tximport` 

***

## Quality control

You have working groups, right? 

```{r set-options, echo=FALSE, cache=FALSE}

out <- data.frame(group = 1:16, condition = rep(c('LM','SA','WL','WR'), each = 4), bio_repl = rep(c(1:4),4))
out$path_to_reads <- paste('/scratch/cluster/monthly/kjaron/sag/RNAseq',out$con, paste(out$con,out$bio_repl,'/', sep = ''), sep = '/')
suppressWarnings(library(knitr))
kable(out)
```

0. Log to Vital-it
1. Go to your working folder (`/scrarch/cluster/monthly/<username>`) and create an output directory for results of quality control (use meaningful name, for instance `<condition><bio_rep>_qc` could be a good idea)

`fastqc` computes a quality control and outputs it to existing output folder defined by parameter `-o`. 

```bash
fastqc –o <fastqc_results> <sequencing_reads>
```

You can put several fastq files with empty space between them. Run fastqc on your files.

2. Submit a job, computing a quality control of the biological replicate of yours.

```bash
bsub –n 2 –J <job_name> fastqc –t 2 -o <output_folder> <fastq_files>
```

3. Check `bjobs` to see status of your job and email if you had a problem.

<sub> For those, who have following error `-bash: fastqc: command not found`. You probably messed up something last time, when we were preparing `~/.bashrc`. Right now, just execute command `module add UHTS/Quality_control/fastqc/0.11.2`, rerun the `fastqc` job and then ask somebody to fix the problem with the setting of `~/.bashrc`. </sub>

4. Copy all fastqc results to your computer
5. Open the files `<fastqc_report>.html` and check and interpret the different graphs. How many reads do you have? Are the reads single or paired-end? Do all the reads have a good quality? Is there any variance in quality or in the number of reads between technical replicates? 
6. Compare your output with biological replicate made other groups.

***

Credits: made by Andrea and Kamil