#Unix

During this semester, we will perform analyses requiring a basic knowledge of bash terminal. If you do not feel, that you are familiar with it, you probably want to check unix_refresh.html. Still here? Good, read please following section about syntax which will be common in all sessions we will have this semester.


## Syntax of this exersize

* blocks of code (bash or R) will be in boxes, note also the usual syntax of Unix commands `command -parameters location`. Comments in code are after `#` symbol.

```bash
ls -h / # list files and directories in -human readable format located at root directory
```

* The comment symbol of Bash is `#`, everything after this symbol is ignored 
* Anything in `< >` you have to replace by meaningful paths, file names or parameters.

```bash
ssh <username>@prd.vital-it.ch #this will be written in this document
ssh kjaron@prd.vital-it.ch #this is example, how it can be used
```

* Bash commands are meaningful acronyms, the explanation is written after every first occurrence in brackets.
* Detail explanations will be written in blocks as follows

<sub>These explanatory blocks should help you to get oriented in documents of Sequence a genome class and to give you an idea how to read it. So, every time you feel confused, be read this blocks and if it does not help, ask one of the assistants.</sub>

***

## Check your wokring folder

1. connect to Vital-it
2. go to your home directory `/home/<user_name>` and check the content using `ls` (list segments).
3. go to your working directory. If it does not exist for any reason, make the directory using `mkdir <dir_name>` (make directory). Check the content. What do you expect?

```bash
mkdir /scratch/cluster/monthly/mls_2015/<username> #if it doesnt exist
cd /scratch/cluster/monthly/mls_2015/<username>
ls -l
```

it is probably empty. `scratch/monthly/` is a working place, all files and folders older than one month are automatically deleted (all data form the last semester).


***

## get the genome, locate the reads

1. go to your working directory (`/scratch/cluster/monthly/mls_2015/<username>`).
2. copy directory `/scratch/cluster/monthly/kjaron/genome` to your working directory using command `cp` (copy)

```bash
cp -r /scratch/cluster/monthly/kjaron/sag/genome .
```

3. Go inside and check the content (`ls`). You should see genome in `.fasta` format, genome anotation in `.gdb` and `.???` formats.

***

## explore those files

Some command line tools for text processing:

- Commands `head -n <num_of_lines> <text_file>`, `tail -n <num_of_lines> <text_file>` and print the beginning, the end of input files, respectively.
- Program `less <text_file>` will allow to browse text file (of any size). You can use scroll or arrows to browse the file. If you press `q`, you will escape back to terminal. If you want to explore, `less` is very good tool.
- Command `wc -cwl` (word count) is counting number of characters, words and lines in the input file. You can specify output by parameters.
- Command `tr` (translate characters) can (not only) delete occurances of string.

1. Formats `.fasta` and `.gds` are just formated plain text formats, so you could use all mentioned command line tools for parsing text. Try it, what is inside of the `.gds`? 

2. How many nucleotides are aproximately in the genome? File `Pseud_protegens_S5_final.fasta` is a simple fasta file, so number of nucleotides is equal to number of all no newline characters in the file without the first line (header). So, if we ignore the header of the fasta you can just use `wc -c <file>`.
3. Since, the most of bash command line tools are very simple, you can easily combine them. The output of commands is usually printed to terminal (list of files in case of `ls`, head of a file in case of `head`, etc.). Using symbol `|` you can redirect ("pipe") an output as the input to an another command. We can use a few commands to find out how many nucleotides are precisely in the genome of *Pseudomonas protegens* have sequenced last semester.

If the argument `-n` of command `tail` is in format `+<first_line_to_print>`, it prints all lines starting by the `<first_line_to_print>`th one, to the end of file. To ensure, that in the file are no newline characters `\n`, we can use `tr -d <file>`, in the end we will use the came command for computing hte number of characters.

```bash
tail -n +2 <file> | tr -d '\n' | wc -c
```

So, how big the genome is? Do you remember approx the size of the *P. protegens* genome from the last semester? Is it consistent with your knowledge?

<sub>
The sequences in `.fasta` files are usually stored in 60 nucleotides per line, but bioinformatians are messy and the format of the `.fasta` slightly differs sometimes. In the case of our genome there are no newline characters in the sequence (i.e. the whole gneome is in the second line of the `.fasta` file), you can check this by deleting `| tr -d '\n'` from the command. Note, that amost all softwares are automatically wrapping too long line to make a document more readible. Using `less Pseud_protegens_S5_final.fasta` can give you a feeling of a usuall `.fasta` even we already know, that is not the case. Also, if our assembly would have several contigs, there should be more header lines in the files, so you would have to use different trick to filter nucleotide sequences only..
</sub>

<sub>
There are many ways how to read or edit plain text files using command line tools. Next to the presented commands, there are several scripting languages specialised for text processing like `awk` or `sed`. In bioinformatics is also common to use `biopython` or `BioPerl` to work with `.fasta` and `.fastq` files.
</sub>

***

time for a 

***

## extract cds

***

## Vital-it

1. `fastqc` computes a quality control and outputs it to existing output folder defined by parameter `-o`. 

```bash
fastqc â€“o <fastqc_results> <sequencing_reads>
```

5. Now you should be able to create an output folder `<fastqc_results>` and submit a job computing a quality control of read file `LM2_L1_R1.fastq`.
6. Check the status of the submitted job
7. Once it finishes, verify existence of files. Copy those files to your computer (explained at beginning) and check the html file.

***

## Challenges

1. Extract all header lines of `LM2_L1_R1.fastq` file to separated file. If you do the same on `Pseud_protegens_S5_final.fasta`, how many headers you extract?

```{bash echo=FALSE, eval=FALSE}
awk '/>/{ print }' LM2_L1_R1.fastq > LM2_L1_R1_headers.txt
wc -l LM2_L1_R1_headers.txt
```

2. Find a way, how to compute total number of nucleotides in `LM2_L1_R1.fastq`.
3. Command `touch` will modify a time of last modification of a file. Find a way, how to touch all files in `/scratch/cluster/monthly` to ensure, that your data will stay there for another month. Note, that writing a script doing this automatically is against rules of Vital-it.

<sub>There are loads of different solutions, some possibilities are written in the source file (`unix_refresh.Rmd`) of this `.html` document. <sub\>

***

# Install needed software

You will need `R` and `IGV` on your computer and programs `fastqc` and `Kallisto` prepared to use on Vital-it or on your computer (OS X, Linux). Make your mind if you want to run programs locally or on the cluster and follow one of the guides.

- `R` - is expected to be already installed on your computer.
- `IGV` - is literally Integrative Genomics Viewer. Follow instructions on the [webpage](https://www.broadinstitute.org/software/igv/download#binary) to install `IGV`.
- `fastqc` - is computing statistics of ngs reads for quality control, specifics bellow.
- `Kallisto` - is quantifying abundances of transcripts from RNA-seq data, specifics bellow.
- `samtools`


### OS X

`Kallisto` - A different ways how to get it can be found on [their webpage](https://pachterlab.github.io/kallisto/starting.html). We will install Kallisto using homebrew - package manager for OS X. Homebrew you can install using following command

```bash
ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
```

<sub> command `curl` downloads an installation script from the URL and ruby is a scripting language which will execute downloaded script. (Ruby is some ways similar with Pearl or Python). Analogical command to `curl` on Linux is `wget`.</sub>

than switch to scientific collection and install kallisto

```bash
brew tap homebrew/science
brew install kallisto 
```

`fastqc` is also included in scientific repository of homebrew

```bash
brew install fastqc
```

### Linux

`Kallisto` for Linux have to be build from source files, no worries, it sounds harder that it is. Use [building manual](https://pachterlab.github.io/kallisto/source.html) and if you get stack or if you have any question, do not hesitate, ask.
`fastqc` is written in Java, check if you have installed Java v1.6 or higher. 
```bash
java -version
```
if is it fine, download fastqc from [Babraham bioinformatics webpage](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/). Unzip the file. Go to the folder, change permissions to make `fastqc` executable and link it to `/usr/bin/`.
```bash
chmod u+x fastqc
sudo ln -s /path/to/FastQC/fastqc /usr/local/bin/fastqc
```

<sub>as you might know, if you are using Linux, `sudo` will execute command with administrator rights, therefore the command will ask you for a password to your account. Also, if your account has no administrator rights you wont be able to link the executable to your `/usr/bin/`. In this case you can add path to `fastqc` to your PATH variable as it is described in vital-it section or you can just run it using full path to executive.</sub>

### Vital-it

Log to vital-it. Open file `/home/<your_user_name>/.bashrc` in `vi` or `nano`. Add there lines:

```bash
module add UHTS/Analysis/kallisto/0.42.4;
module add UHTS/Quality_control/fastqc/0.11.2;
module add UHTS/Analysis/samtools/latest
```

<sub> script `.bashrc` located at your home directory is a bash script which is executed every time you start a new session of terminal. You can easily check how it works by adding there a line `echo I love sequence a genome class`. Every time, you will log to Vital-it you will see this important message. The dot at the begging of the file name means, that it is hidden file, therefore if you go to your home `cd ~` and list your files there `ls` you wont see it. You have to use `ls -a` option, which stands for all. In file starting by `.` is usually stored a local user setting of some program, meta files or history (like `.R` settings of R, `bash_history` bash history, `.ssh` setting of SSH, etc.). </sub>

***

## Install needed R packages

### any OS

Start a R on your computer (NOT on vital-it) and install following packages from bioconductor:

* `edgeR` on [webpage](https://bioconductor.org/packages/release/bioc/html/edgeR.html)
* `tximport` 

***

## Tips and tricks

- Command `cd` without any argument changes the working directory to your home. Symbol `~` works the same, it is symbol of home.
- Press tabulator to call whisperer every time you are writing a path. It is good way how to prevent typos.
- Arrows up and down allow to browse your command history saved at `~/.bash_history`
- Press `Ctrl+R` for a full text search in the command history
- Press `Ctrl+C` to stop the running process, useful `cat` something, what you should not...
- You can use wild characters `?`, `[]` and `*`, to specify a group of files. `?` stands for any single character, `[XYZ]` one of the specifies characters (XYZ in this case, you can also use numbers), `*` any number of any characters. See examples

```bash
cat *.txt # prints all text files
ls ?? # lists all files or directories with exactly two letter names
cp /<path>/LS[1234]* . # copiing all files staring by LS1 LS2 LS3 or LS4 from given <path>
```

- Bash script `~/.bashrc` is executed every time you open new window terminal.
- (Linux only) You can make your command prompt colourful by adding following line to `~/.bashrc`

```bash
PS1='${debian_chroot:+($debian_chroot)}\[\033[01;32m\]\u@\h\[\033[00m\]:\[\033[01;34m\]\w\[\033[00m\]\$ '
```

***

Credits: made by Andrea and Kamil