---
output: html_document
---
#Unix

During this semester, we will perform analyses requiring a basic knowledge of bash terminal. If you do not feel, that you are familiar with it, you probably want to check `unix_refresh.html`. Still here? Good, read please following section about syntax which will be common in all sessions we will have this semester.

## Syntax of this exercise

* Blocks of code (bash or R) will be in boxes, note also the usual syntax of Unix commands `command -parameters location`. Comments in code are after `#` symbol.

```bash
ls -h / # list files and directories in -human readable format located at root directory
```

* The comment symbol of Bash is `#`, everything after this symbol is ignored 
* Anything in `< >` you have to replace by meaningful paths, file names or parameters.

```bash
ssh <username>@prd.vital-it.ch #this will be written in this document
ssh kjaron@prd.vital-it.ch #this is example, how it can be used
```

* Bash commands are meaningful acronyms, the explanation is written after every first occurrence in brackets.
* Detail explanations will be written in blocks as follows

<sub>These explanatory blocks should help you to get oriented in documents of Sequence a Genome class and to give you an idea how to read it. So, every time you feel confused, please read this blocks and if it does not help, ask one of the assistants.</sub>

***

### OS X, Linux users

**connection**

Open terminal and connect to Vital-it using `ssh` (secure shell)
```bash
ssh <username>@prd.vital-it.ch
```

**file transfer**

To copy file **to** cluster you have to be in the terminal of your computer, then
```bash
scp <path>/<filename> <username>@prd.vital-it.ch:/<path>/<filename> # (scp stands for secure copy)
```

to copy file **from** the cluster to your computer, you still have to be in the terminal of your computer, but the order of locations in the command will be just reverse
```bash
scp <username>@prd.vital-it.ch:/<path>/<filename> <path>/<filename>
```

***

### Windows users

**connection**

1. Install [WinSCP](http://winscp.net/eng/download.php) and [Putty](http://www.chiark.greenend.org.uk/~sgtatham/putty/download.html)
2. First start WinSCP by:
    -	hostname: prd.vital-it.ch
    -	username: 
    -	password:
3. Then start your Putty entering hostname prd.vital-it.ch

**file transfer**

Transfer the file to your computer by drag and drop from the server side. Every time in the exercise you see `scp` command, it is only for Mac/Linux users, you use WinSCP. In the WinSCP navigate your paths and check your folders. 


***

## Check your working folder

1. connect to Vital-it
2. go to your home directory `/home/<user_name>` and check the content using `ls` (list segments).
3. go to your working directory. If it does not exist for any reason, make the directory using `mkdir <dir_name>` (make directory). Check the content. What do you expect?

```bash
mkdir /scratch/cluster/monthly/mls_2015/<username> #if it doesnt exist
cd /scratch/cluster/monthly/mls_2015/<username>
ls -l
```

it is probably empty. `scratch/monthly/` is a working place, all files and folders older than one month are automatically deleted (all data from the last semester).


***

## Get the genome, locate the reads

1. go to your working directory (`/scratch/cluster/monthly/mls_2015/<username>`).
2. copy directory `/scratch/cluster/monthly/kjaron/genome` to your working directory using command `cp` (copy)

```bash
cp -r /scratch/cluster/monthly/mls_2015/SAGE/genome .
```

3. Go inside and check the content (`ls`). You should see genome in `.fasta` format, genome annotation in `.gbk` and `.gtf` formats.

All RNA-seq files, are saved in `/scratch/cluster/monthly/mls_2015/SAGE/RNAseq/<condition>/<biological_replicate>.fastq.gz`. 

4. Go to the folder, check the content. How many conditions we have got? To have overview of sizes of all folders and the sum of sizes of their content, you can use `du -h`.
5. Exploring further, how many biological replicates per condition we have got? 

<sub>
Genome files are relatively small, therefore it is fine, if every one of you will have a copy of the file in your folder. However, the raw fastq reads are 34GB, so do not copy them to your folder, rather use the full path `/scratch/cluster/monthly/mls_2015/SAGE/RNAseq/...` to the files, we will share.
</sub>

***

## Explore those files

Some command line tools for text processing:

- Commands `head -n <num_of_lines> <text_file>`, `tail -n <num_of_lines> <text_file>` and print the beginning (first n lines), the end of input files (last n lines), respectively.
- Command `less <text_file>` will allow to browse text file (of any size). You can use scroll or use arrows to browse the file. If you press `q`, you will escape back to the terminal. If you want to explore, `less` is very good tool.
- Command `wc -lwc <text_file>` (word count) is counting number of newline characters, words and characters in the input file. You can specify output by parameters .
- Command `tr -d '<string_to_be_removed>' <text_file>` (translate characters) can delete all occurrences of a given string (Note: a character is also a string).

<sub>
The number of lines can differ from the number of newline characters, if the last line of the file is without newline character in the end.
<sub\>

1. Formats `.fasta` and `.gbk` are just formatted plain text formats, so you could use all mentioned command line tools for parsing text. Try it, what is inside of the `.gbk`? 

2. How many nucleotides are approximately in the genome? File `Pseud_protegens_S5_final.fasta` is a simple fasta file, so the number of nucleotides is equal to the number of all characters other than newline in the file without the first line (header). So, if we ignore the header of the fasta you can just use `wc -c <file>`.
3. Since most of bash command line tools are very simple, you can easily combine them. The output of commands is usually printed to terminal (list of files in case of `ls`, head of a file in case of `head`, etc.). Using the symbol `|` you can redirect ("pipe") an output as the input to an another command. We can use a few commands to find out how many nucleotides are precisely in the genome of *Pseudomonas protegens* we have sequenced last semester.

If the argument `-n` of command `tail` is in format `+<first_line_to_print>`, it prints all lines starting by the `<first_line_to_print>`th one, to the end of file. To ensure that in the file are no newline characters `\n`, we can use `tr -d <file> '\n'`, in the end we will use the same command for computing the number of characters.

```bash
tail -n +2 <file> | tr -d '\n' | wc -c
```

So, how big is the genome? Do you remember approximatively the size of the *P. protegens* genome from last semester? Is it consistent with your knowledge?

<sub>
The sequences in `.fasta` files are usually stored in 60 nucleotides per line, but bioinformaticians are messy and the format of the `.fasta` slightly differs sometimes. In the case of our genome there are no newline characters in the sequence (i.e. the whole genome is in the second line of the `.fasta` file), you can check this by deleting `| tr -d '\n'` from the command. Note that almost all software are automatically wrapping too long lines to make a document more readable. Using `less Pseud_protegens_S5_final.fasta` can give you a feeling of a usual `.fasta` even if we already know that is not the case. Also, if our assembly would have several contigs, there should be more header lines in the files, so you would have to use different trick to filter nucleotide sequences only..
</sub>

<sub>
There are many ways how to read or edit plain text files using command line tools. Next to the presented commands, there are several scripting languages specialised for text processing like `awk` or `sed` or `grep`. In bioinformatics it is also common to use `biopython` or `BioPerl` to work with `.fasta` and `.fastq` files.
</sub>

***

## Break!

Now we know what we have, so it is a good time to stop typing commands and to start to ask... What are we doing? What do we want to do? What is the meaning of life?

***

## submit a quality control job

You have working groups, right? 

```{r set-options, echo=FALSE, cache=FALSE}

out <- data.frame(group = 1:16, condition = rep(c('LM','SA','WL','WR'), each = 4), bio_repl = rep(c(1:4),4))
out$path_to_reads <- paste('/scratch/cluster/monthly/mls_2015/SAGE/RNAseq',out$con, paste(out$con,out$bio_repl,'.fastq.gz', sep = ''), sep = '/')
suppressWarnings(library(knitr))
kable(out)
```

1. Log to Vital-it.
2. Activate Vital-it module `fastqc`.

```bash
module add UHTS/Quality_control/fastqc/0.11.2;
```

3. Go to your working folder (`/scratch/cluster/monthly/mls_2015/<username>`) and create an output directory for results of quality control (use meaningful name, for instance `<condition><bio_rep>_qc` could be a good idea).

`fastqc` computes a quality control and outputs it to existing output folder defined by parameter `-o`. 

```bash
fastqc -o <output_folder> <sequencing_reads.fastq.gz>
```

You can put several fastq files with empty space between them. Run fastqc on your files.

4. Submit a job, computing a quality control of the biological replicate of yours.

```bash
bsub -n 2 -J <job_name> fastqc -t 2 -o <output_folder> <sequencing_reads.fastq.gz>
```

5. Check `bjobs` to see status of your job and email if you had a problem (when job is finished, you will get an email).

<sub> For those, who have following error `-bash: fastqc: command not found`. You probably messed up something last time, when we were preparing `~/.bashrc`. Right now, just execute command `module add UHTS/Quality_control/fastqc/0.11.2`, rerun the `fastqc` job and then ask somebody to fix the problem with the setting of `~/.bashrc`. </sub>

***

## Extract cds

1. go to `<your_working_directory>/genome/`
2. activate Vital-it module `cufflinks`

```bash
module add UHTS/Assembler/cufflinks/2.2.1;
```

3. use `gffread` from cufflinks to extract coding regions from `.fasta` and `.gtf` files

```bash
gffread  -g <S5_genome.fasta> -x <output.fasta> <S5_annotation.gtf>
```

<sub>
`<output.fasta>` is just any name you will give to your newly created file with coding regions. Wise idea could be to put into name an organism, strain and cds. The `gffread` will also output the index file, which we wont use.
<sub\>

4. Check the file `<output.fasta>` using `less` or `head`. Does it look like extracted coding regions? 
5. How many transcripts do we have?

```bash
awk '/>/{ print }' <output.fasta> | wc -l
```

<sub>
`awk` is a scripting language for parsing text files. What is in curly brackets `'{ command }'` is executed on every line. If there is a pattern written in slashes before, the command will be executed only on lines with this pattern. In our case, the command print whole line (`print`) is executed on all lines containing symbol `>`, which represent lines with headers. Counting number of headers will give us a number of cds we extracted.
</sub>

***

## Read a quality control output

1. Copy all fastqc results to your computer.
2. Open the files `<fastqc_report>.html` and check and interpret the different graphs. How many reads do you have? Are the reads single or paired-end? Do all the reads have a good quality? Is there any variance in quality or in the number of reads between technical replicates? 
3. Are there any red flags?? Is it a problem?

<sub>
The `fastqc` is a program designed for genomic reads in first place, therefore some red flags are not crucial for us. For instance funky QC content or kmer bias are actually expectable properties of RNA seqeunces. What is important is to have high quality data, no adapter content and not too short reads.
<sub\>

***

## Adapter trimming and quality filtering

1. **Bonus**: Searching for adapters (they are actually known). If you dig into source code of fasqc, or check Illumina website, you can find seqeunces of adapters and usually you find an adapter matching your adapter contamination. Another aproach would be to check kmer frequencies, but this is not that practical for RNAseq. We found that adapter of `LM2.fastq.gz` contains sequence `ATCTCGTATGCCGTCTTCTGCTTG`. To identify whole adapter sequence, you can select fulltext search in some subset of your reads (100000 lines). In the subset you use a fulltext search to find reads containing our adapter sequence, select first one hundred of them and show them using `less`.

```bash
zcat /scratch/cluster/monthly/mls_2015/SAGE/RNAseq/LM/LM2.fastq.gz | head -n 100000 | grep ATCTCGTATGCCGTCTTCTGCTTG | head -n 100 | less
```

Command `less` allow you a fulltext search as well. Press `?` and copy-paste there a sequence `ATCTCGTATGCCGTCTTCTGCTTG`, hit `Enter`. Now all `ATCTCGTATGCCGTCTTCTGCTTG` should be highlighted. Can you find a pattern? Do we know whole seqeuence of adapter or is it actually longer?

<sub>
We will actually cut all possible adapters without specifying which one is in our sequence. Therefore we wont use the output of our little adapter investigaton. 
<sub\>

2. Activate Vital-it module `trimmomatic`
```bash
module add UHTS/Analysis/trimmomatic/0.33
```

3. Submit a trimming job (this will take some hours)
```bash
bsub -n 8 -M 8388608 -J <job_name> trimmomatic SE -threads 8 <input_reads.fastq.gz> <output_trimmed.fastq.gz> ILLUMINACLIP:<adapters.fasta>:3:25:6 LEADING:9 TRAILING:9 SLIDINGWINDOW:4:15 MINLEN:60
```

<sub>
Sometimes, when the fragment is really small, there can be more consecutive adapters sequenced. Those adapters are not cut in demultiplexing in the sequencing centre. It is up on us, to get rid of them. The most of possible adapters of Illumina RNAseq are in the file `/scratch/cluster/monthly/mls_2015/SAGE/RNAseq/RNAseq_adapters.fa`. `trimmomatic` cat try to identify those sequences in the reads and cut them off if the module `ILLUMINACLIP` is used. You have to specify the file with adapters. Other modules of trimmomatic, `LEADING`, `TRAILING` are cutting given number bases from the start and end of read if the quality is below some internal threshold. Module `SLIDINGWINDOW` performs a sliding window trimming, cutting once the average quality within the window falls below a threshold. Module `MINLEN` simply discards all reads which are after all that trimming shorter than given number of bases.
<sub\>

***

# Install needed software (homework)

## Compulsory part

You will need `R` and `IGV` on your computer and two R packages `edgeR` and `tximport`.

- `R` - is expected to be already installed on your computer.
- `IGV` - stands for Integrative Genomics Viewer. Follow instructions on the [webpage](https://www.broadinstitute.org/software/igv/download#binary) to install `IGV`.

Start a R session on your computer (NOT on vital-it) and install the following packages from bioconductor:

* `edgeR` on [webpage](https://bioconductor.org/packages/release/bioc/html/edgeR.html)
* `tximport` 

<sub>
If you are not sure how to install bioconductor packages, visit [bioconductor webpage](https://www.bioconductor.org/install/)
<sub\>

## Optional part

Nobody likes it, but a lot of time of bioinformaticians is consumed by software installation. If you want to try run programs `fastqc`, `Kallisto` and `samtools` on your computer (OS X, Linux) to get training in this not that favourite discipline, install those programs on your computer using following instructions. This software is easily installable for Linux and OS X users, the Windows users have to compute everything on Vital-it or face an extreme challenge.

- `fastqc` - is computing statistics of ngs reads for quality control, specifics bellow.
- `Kallisto` - is quantifying abundances of transcripts from RNA-seq data, specifics bellow.
- `samtools` - set of tools for manipulation with alignment formats `.sam` and `.bam`

### OS X

`Kallisto` - A different ways how to get it can be found on [their webpage](https://pachterlab.github.io/kallisto/starting.html). We will install Kallisto using homebrew - package manager for OS X. Homebrew you can install using following command

```bash
ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
```

<sub> command `curl` downloads an installation script from the URL and ruby is a scripting language which will execute downloaded script. (Ruby is some ways similar to Perl or Python). Analogical command to `curl` on Linux is `wget`.</sub>

than switch to scientific collection and install kallisto

```bash
brew tap homebrew/science
brew install kallisto 
```

`fastqc` is also included in scientific repository of homebrew

```bash
brew install fastqc
```

### Linux

`Kallisto` for Linux have to be build from source files, no worries, it sounds harder than it is. Use [building manual](https://pachterlab.github.io/kallisto/source.html) and if you get stuck or if you have any question, do not hesitate to ask.
`fastqc` is written in Java, check if you have installed Java v1.6 or higher. 
```bash
java -version
```
if is it fine, download fastqc from [Babraham bioinformatics webpage](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/). Unzip the file. Go to the folder, change permissions to make `fastqc` executable and link it to `/usr/bin/`.
```bash
chmod u+x fastqc
sudo ln -s /path/to/FastQC/fastqc /usr/local/bin/fastqc
```

<sub>as you might know, if you are using Linux, `sudo` will execute command with administrator rights, therefore the command will ask you for a password to your account. Also, if your account has no administrator rights you wont be able to link the executable to your `/usr/bin/`. In this case you can add path to `fastqc` to your PATH variable as it is described in vital-it section or you can just run it using full path to executive.</sub>

***

Credits: made by Andrea and Kamil with kind help of Amina and Olivier. Contact: 376090@mail.muni.cz
