---
output: html_document
---
#Unix refreshement

During this semester, we will perform analyses requiring a basic knowledge of Unix terminal. If you feel, that you are familiar with it, you can try to do exercises in the section Challenges. If you find them easy, feel free not to the previous exercises. However, please read following section about syntax which will be common in all sessions we will have this semester.

Here are more materials about [UNIX](http://edu.isb-sib.ch/course/view.php?id=82), [R](http://edu.isb-sib.ch/course/view.php?id=52) and [Vital-it](http://edu.isb-sib.ch/course/view.php?id=72).

## Syntax of this exersize

* blocks of code (bash or R) will be in boxes, note also the usual syntax of Unix commands `command -parameters location`. Comments in code are after `#` symbol.

```bash
ls -h / # list files and directories in -human readable format located at root directory
```

* Anything in `< >` you have to replace by meaningful paths, file names or parameters.

```bash
ssh <username>@prd.vital-it.ch #this will be written in this document
ssh kjaron@prd.vital-it.ch #this is example, how it can be used
```

* Detail explanations will be written in blocks as follows

<sub>These explanatory blocks should help you to get oriented in documents of Sequence a genome class and to give you an idea how to read it. So, every time you feel confused, be read this blocks and if it does not help, ask one of the assistants.</sub>

---
### OS X, Linux users

**connection**

Open terminal and connect to Vital-it
```bash
ssh <username>@prd.vital-it.ch
```

**file transfer**

To copy file to cluster, you have to be in the terminal of your computer, then
```bash
scp <path>/<filename> <username>@prd.vital-it.ch:/<path>/<filename>
```

to copy file from the cluster to your computer, you still have to be in the terminal of your computer, but the order of locations in the command will be just reverse
```bash
scp <username>@prd.vital-it.ch:/<path>/<filename> <path>/<filename>
```


### Windows users

**connection**

1. Install [WinSCP](http://winscp.net/eng/download.php) and [Putty](http://www.chiark.greenend.org.uk/~sgtatham/putty/download.html)
2. First start WinSCP by:
    -	hostname: prd.vital-it.ch
    -	username: 
    -	password:
3. Then start your Putty entering hostname prd.vital-it.ch

**file transfer**

Transfer the file to your computer by drag and drop from the server side. Every time in the exercise you see `scp` command, it is only for Mac/Linux users, you use WinSCP. In the WinSCP navigate your paths and check your folders. 

***

## Warm-up

0. connect to Vital-it
1. go to your home directory `/home/<user_name>` and check the content (`ls`).
2. go to your scratch directory. If it does not exist for any reason, make the directory using `mkdir <dir_name>`. Check the content. What do you expect?

```bash
mkdir /scratch/cluster/monthly/mls_2015/<username> #if it doesnt exist
cd /scratch/cluster/monthly/mls_2015/<username>
ls -l
```

it is probably empty. `scratch/monthly/` is a working place, all files and folders older than one month are automatically deleted (all data form the last semester).

3. make a directory
4. Use `pwd` to check your location. Go up one directory `cd ..` and verify the success using `pwd` command again. Generally, `cd` changes a working directory to the provided path `cd <path>`.
5. Try command `man <command>` to show manual page of some commands. What means the `-l` in the `ls` command few line above?

<sub>The command `cd` and correct handling of paths is essential for everything done is Unix terminal. The path can be defined relatively to your current directory (ex: `cd ./data` or `cd ..`) or absolutely, i.e. starting from the root directory of the computer (ex: `cd /scratch/cluster`). Meanings of symbols are as follows: `.` = current directory, `..` = directory one lever above current directory, `/` = root directory, `~` = your home directory. Note that symbol `~` carry different path for every user: `/home/<username>`</sub>

***

## Copy, move, rename and remove

1. go to your working directory (`/scratch/cluster/monthly/mls_2015/<username>`).
2. copy directory `/scratch/cluster/monthly/kjaron/unix_ex` to your working directory

```bash
cp -<parameters> <path/file> <path/new_file_name> #generally for copy of files
# if you dont provide a new name, the old one will be kept
cp -r /scratch/cluster/monthly/kjaron/sag/unix_ex . #directories need to be copied using -r
```

3. Go inside and check the content (you should know how, if you don't, scroll up).
4. Copy file `text_file.txt` in directory `dir_1`. Do you need parameter `-r`? If you not sure, check manual page.
5. Go to `dir_1`, rename `text_file.txt` to `<your_new_name>.txt` and move it back to `unix_exec`.

```bash
cd ./dir_1
mv text_file.txt <your_new_name>.txt
mv <your_new_name>.txt ..
cd ..
```

6. What is the difference between second and the third line? Can you figure out how to do the same without changing the directory to `dir_1`? 

<sub>Command `mv` moves file to the given path and rename it to the given name. So, if the `<path_to_input>` in the command `mv <path_to_input>/<file_name> <path_to_output>/<new_file_name>` is the same as `<path_to_output>`, the file will be only renamed. If the path will be different, but `<new_file_name>` wont be provided, the file will be just moved (not renamed).</sub>

7. Remove the copied file.

```bash
rm <your_new_name>
```

8. Remove directory `dir_2`. Problem? Check manual or Google.
9. **bonus** copy again file `text_file.txt` in directory `dir_1`. Can you move the copy to your directory and rename it using one command only?

***

## Text processing

There are many many ways how to show a plain text file in the terminal, so if you find another solutions than we provide, do not wonder.

1. Use commands `head -n <num_of_lines> <text_file>`, `tail -n <num_of_lines> <text_file>` and `cat <text_file>` to see the beginning, the end or whole file `text_file.txt`, respectively. How many lines are in the file?
2. Use `less <text_file>` to check `hox_bflo.fa`. You can use scroll or arrows to move in the file. If you press `q`, you will escape back to terminal. You can also use `head` and `tail` commands to see the beginning and the end of the file. Do you have an idea, what would happen if you would use `cat`?
3. To get an idea, about the file, you can use command `wc`, which is counting number of characters, words and lines in the input file. You can specify output by parameters. How many lines are in `text_file.txt`?
4. Standard output is usually printed to terminal (list of files in case of `ls`, content of file in case of `cat`, etc.). Using symbol `|` you can "pipe" the standard output as input of another command. We are interested in number of nucleotides in Amphioxus HOX cluster, which is equal to number of all no newline characters in the `hox_bflo.fa` file without the first line (header).

```bash
head -n 1 hox_bflo.fa | wc -c
cat hox_bflo.fa | wc -cl
```

5. You can simplify your problem using command `tail` which can also print all but a given number of lines from begging. Following command prints number of characters (including newline characters) and number of newlines. What is the size of Amphioxus' HOX cluster? 

```bash
tail -n +2 hox_bflo.fa | wc -cl
```

6. **bonus** if you would like to simplify your life even more, you can use `tr` command to delete all newline characters before you count them using `wc`. Note, that commands can be nested, so you can use as many pipes, as you need.

```bash
tail -n +2 hox_bflo.fa | tr -d '\n' | wc -c
```

<sub>
There are many ways how to read or edit plain text files using command line tools. Next to the presented commands, there are several scripting languages specialised for text processing like `awk` or `sed`. In bioinformatics is also common to use `biopython` or `BioPerl` to work with fasta and fastq files.
</sub>

7. Editing files can be done using editors `nano <textfile>` or `vi <textfile>`. Open `text_file.txt` in `vi`, press `a` to ender to editing mode. Add number 11 to the series of number. Press `<Esc>` to escape from the writing mode and `:` to enter to command mode, write `wq` to save changes (write) and exit back to terminal (quit). Verify, that you managed to add there the value by `cat`, `less` or using `wc` (should return 20 lines now).

<sub> The whole point of two modes (command and editing mode) of `vi` is that command mode allows you to access advanced functions like find and replace, jump to particular line or deleting several lines, which makes `vi` very powerful editor, even it has no graphical user interface. `vi` is common part of Unix terminal and therefore you should be able to at least make minor edits and save changes, if there is no any other text editor installed, `vi` will be. If you want to quit `vi` without saving, you have to write exclamation mark `:q!`. The `:q` can quit `vi` only if the opened file remain unchanged. It is also worth to note, that command `vi` is usually linked to program `vim`, what is more user friendly generation of the editor (you do not want to know how "friendly" was the original `vi`).</sub>

***

## Vital-it

1. You will need two main commands: `bsub` to submit a job and `bjobs` to obtain job status. Submission of a job have following syntax 

```bash
bsub <bsub_parameters> <command> <command_parameters>
```

<sub>
Vital-it is a computing cluster - hundreds of computers connected in a huge grid. The computer you are logging to, when you want to submit a job (prd.vital-it.ch) is one computer (server) and it serves exclusively for distribution of computational tasks accross computing grid. This computer is called front-end. Everytime, you run a `bsub` on the front-end, it will put your job in the queue and assign you a computer once, there will be one free for you, given specifications you have provided in `<bsub_parameters>`. 
</sub>

2. Programs available on the Vital-it you can loaded using 

```bash
module add <field>/<category>/<program>/<version>
```

3. The list of programs and commands for loading them can be found on [Vital-it webpage](http://www.vital-it.ch/software/tools.php). Find there a command for program fastqc and run it.
4. `fastqc` computes a quality control and outputs it to existing output folder defined by parameter `-o`. 

```bash
fastqc <sequencing_reads> –o <fastqc_results>
```

5. Now you should be able to create an output folder `<fastqc_results>` and submit a job computing a quality control of read file `NEED TO BE ADDED`.
6. Check the status of the submitted job
7. Once it finishes, verify existence of files. Copy those files to your computer (explained at beginning) and check the html file.

***

## Challenges

1. Extract all header lines of `NEED TO BE ADDED` file to separated file. If you do the same on `hox_bflo.fa`, how many headers you extract?
2. Find a way, how to compute total number of nucleotides in `NEED TO BE ADDED`.
3. Command `touch` will modify a time of last modification of a file. Find a way, how to touch all files in `/scratch/cluster/monthly` to ensure, that your data will stay there for another month. Note, that writing a script doing this automatically is against rules of Vital-it.

***

# Install needed software

You will need `R` on your computer and following programs on your computer (OS X, Linux) or prepared to use on vital-it. Make your mind if you want to run programs locally or on the cluster and follow one of the guides.

`fastqc` - is computing statistics of ngs reads for quality control
`Kallisto` - is quantifying abundances of transcripts from RNA-seq data.

### OS X

`Kallisto` - A different ways how to get it can be found on [their webpage](https://pachterlab.github.io/kallisto/starting.html). We will install Kallisto using homebrew - package manager for OS X. Homebrew you can install using following command

```bash
ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
```

<sub> command `curl` downloads an installation script from the url and ruby is a scripting language which will execute downloaded script. (Ruby is some ways similar with Pearl or Python). Analogical command to `curl` on Linux is `wget`.</sub>

than switch to scientific collection and install kallisto

```bash
brew tap homebrew/science
brew install kallisto 
```

`fastqc` is also included in scientific repository of homebrew

```bash
brew install fastqc
```

### Linux

`Kallisto` for Linux have to be build from source files, no worries, it sounds harder that it is. Use [building manual](https://pachterlab.github.io/kallisto/source.html) and if you get stack or if you have any question, do not hesitate, ask.
`fastqc` is written in Java, check if you have installed Java v1.6 or higher. 
```bash
java -version
```
if is it fine, download fastqc from [Babraham bioinformatics webpage](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/). Unzip the file. Go to the folder, change permissions to make `fastqc` executable and link it to `/usr/bin/`.
```bash
chmod u+x fastqc
sudo ln -s /path/to/FastQC/fastqc /usr/local/bin/fastqc
```

<sub>as you might know, if you are using Linux, `sudo` will execute command with administrator rights, therefore the command will ask you for a password to your account. Also, if your account has no administrator rights you wont be able to link the executable to your `/usr/bin/`. In this case you can add path to `fastqc` to your PATH variable as it is described in vital-it section or you can just run it using full path to executive.</sub>

### Vital-it

Log to vital-it. Open file `/home/<your_user_name>/.bashrc` in `vi` or `nano`. Add there lines:

```bash
module add UHTS/Analysis/kallisto/0.42.4;
module add UHTS/Quality_control/fastqc/0.11.2;
```

<sub> script `.bashrc` located at your home directory is a bash script which is executed every time you start a new session of terminal. You can easily check how it works by adding there a line `echo I love sequence a genome class`. Every time, you will log to Vital-it you will see this important message. The dot at the begging of the file name means, that it is hidden file, therefore if you go to your home `cd ~` and list your files there `ls` you wont see it. You have to use `ls -a` option, which stands for all. In file starting by `.` is usually stored a local user setting of some program, meta files or history (like `.R` settings of R, `bash_history` bash history, `.ssh` setting of SSH, etc.). </sub>

***

## Install needed R packages

### any OS

Start a R on your computer (NOT on vital-it) and install following packages from bioconductor:

* `edgeR` on [webpage](https://bioconductor.org/packages/release/bioc/html/edgeR.html)
* `tximport` 

***

## Tips and tricks

- Command `cd` without any argument changes the working directory to your home. Symbol `~` works the same, it is symbol of home.
- Press tabulator to call whisperer every time you are writing a path. It is good way how to prevent typos.
- Arrows up and down allow to browse your command history saved at `~/.bash_history`
- Press `Ctrl+R` for a full text search in the command history
- Press `Ctrl+C` to stop the running process, useful `cat` something, what you should not...
- You can use wild characters `?`, `[]` and `*`, to specify a group of files. `?` stands for any single character, `[XYZ]` one of the specifies characters (XYZ in this case, you can also use numbers), `*` any number of any characters. See examples

```bash
cat *.txt # prints all text files
ls ?? # lists all files or directories with exactly two letter names
cp /<path>/LS[1234]* . # copiing all files staring by LS1 LS2 LS3 or LS4 from given <path>
```

- Bash script `~/.bashrc` is executed every time you open new window terminal.
- You can make your command prompt colourful by adding following line to `~/.bashrc`

```bash
PS1='${debian_chroot:+($debian_chroot)}\[\033[01;32m\]\u@\h\[\033[00m\]:\[\033[01;34m\]\w\[\033[00m\]\$ '
```

***

Credits: based on Nadja's material, made by Andrea and Kamil
